src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+
+            private bool IsUnusedParameterCandidate(IParameterSymbol parameter)
+            {
+                // Ignore implicitly declared parameters and methods, extern methods, abstract methods,
+                // virtual methods, overrides, interface implementations, accessors and lambdas.
+                if (parameter.IsImplicitlyDeclared ||
+                    parameter.Name == "_" ||
+                    !(parameter.ContainingSymbol is IMethodSymbol method) ||
+                    method.IsImplicitlyDeclared ||
+                    method.IsExtern ||
+                    method.IsAbstract ||
+                    method.IsVirtual ||
+                    method.IsOverride ||
+                    !method.ExplicitOrImplicitInterfaceImplementations().IsEmpty ||
+                    method.IsAccessor() ||
+                    method.IsAnonymousFunction())
+                {
+                    return false;
+                }
+
+                // Ignore event handler methods "Handler(object, MyEventArgs)"
+                if (_eventArgsType != null &&
+                    method.Parameters.Length == 2 &&
+                    method.Parameters[0].Type.SpecialType == SpecialType.System_Object &&
+                    method.Parameters[1].Type.InheritsFromOrEquals(_eventArgsType))
+                {
+                    return false;
+                }
+
+                // Ignore methods with any attributes in 'attributeSetForMethodsToIgnore'.
+                if (method.GetAttributes().Any(a => a.AttributeClass != null && _attributeSetForMethodsToIgnore.Contains(a.AttributeClass)))
+                {
+                    return false;
+                }
+
+                // Ignore methods that were used as delegates
+                if (_methodsUsedAsDelegates.ContainsKey(method))
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            private sealed class BlockAnalyzer
+            {
+                private readonly SymbolStartAnalyzer _symbolStartAnalyzer;
+                private readonly Options _options;
+                private bool _hasDelegateCreationOrAnonymousFunction;
+                private bool _escapedDelegate;
+                private bool _hasConversionFromDelegateTypeToNonDelegateType;
+
+                private BlockAnalyzer(SymbolStartAnalyzer symbolStartAnalyzer, Options options)
+                {
+                    _symbolStartAnalyzer = symbolStartAnalyzer;
+                    _options = options;
+                }
+
+                public static void Analyze(OperationBlockStartAnalysisContext context, SymbolStartAnalyzer symbolStartAnalyzer)
+                {
+                    if (HasSyntaxErrors())
+                    {
+                        return;
+                    }
+
+                    // All operation blocks for a symbol belong to the same tree.
+                    var firstBlock = context.OperationBlocks[0];
```

Comment:
is it guaranteed somewhere that thsi list is always non-empty if there are no syntax errors?

Reply:
Yes: [http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerDriver.cs,1925](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerDriver.cs,1925), but it is probably better to be conservative and bail out if not, instead of throw. I have updated the check.

---
In reply to: [230605058](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605058) [](ancestors = 230605058)